### DISCUSSION TITLE
ES6 모듈을 사용했을때, 해당 코드의 결과를 설명해주세요

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#104] [모듈]
- ES6 모듈 (ESM) 을 사용했을때, 해당 코드의 결과를 설명해주세요
```js
if(true){
  import './file.mjs'; 
}

if(true){
  const m1=await import('./file.mjs'); 
}
```

- CommonJS 을 사용했을때, 해당 코드의 결과는?
```jsx
if(true){
	require(./file.js);
}
```

**Comment by publdaze** [comment]
> * ES6 모듈 (ESM) 을 사용했을때, 해당 코드의 결과를 설명해주세요
> 
> ```js
> if(true){
>   import './file.mjs'; 
> }
> 
> if(true){
>   const m1=await import('./file.mjs'); 
> }
> ```
- import 문은 정적 분석을 제공하기 위해 파일(모듈) 최상단에 위치해야 함. 함수나 블록 내부에서 사용 불가. 따라서 첫번째 import 문에서 에러 발생
- 위에서 에러가 발생하면 아래 실행문으로 더이상 진행 안하지만, 별도로 실행한다고 보고,
- ES2022 이전에는 await을 async 없이 사용하면 syntax에러가 발생하지만, ES2022 이후에는 최상위 수준에서는 async 없이 사용가능하기 때문에 정상 실행된다 가정하면,
- import(module) 표현식은 모듈이 export한 객체의 프로미스를 반환하기 때문에 await을 통해 동적으로 import 가능.

> * CommonJS 을 사용했을때, 해당 코드의 결과는?
> 
> ```js
> if(true){
> 	require(./file.js);
> }
> ```

- 모듈 경로는 문자열로 제공해야 하므로 SyntaxError 발생
- 문자열로 되어 있다 가정하면,
- 동적으로 모듈 호출이 가능하여 if 내부에서도 모듈 호출 가능

**Comment by pipisebastian** [comment]
> * ES6 모듈 (ESM) 을 사용했을때, 해당 코드의 결과를 설명해주세요
> 
> ```js
> if(true){
>   import './file.mjs'; 
> }
> 
> if(true){
>   const m1=await import('./file.mjs'); 
> }
> ```


- `1` import 문이 정적이므로 최상위에서만 사용 가능, 아닐 경우 문법 오류가 발생
- `2` 동적 import()는 비동기 함수 내에서만 사용할 수 있으며, 그렇지 않으면 문법 오류가 발생. 비동기 함수 내에서는 정상적으로 작동


> * CommonJS 을 사용했을때, 해당 코드의 결과는?
> 
> ```js
> if(true){
> 	require(./file.js);
> }
> ```

- 조건문 내에서 require를 사용해 모듈을 동기적으로 가져올 수 있음
